### PR TITLE
Additional check for QualificationValue

### PR DESCRIPTION
**Pull request by apetushok**
for #98

### CODE REVIEW

**Comment by apetushok** [comment]
I have verified that both characters are valid and are used to display the date. So this simple check should be enough.

image

**Comment by mzeinstra** [comment]
They are valid components but they are not valid when used together.

See https://www.loc.gov/standards/datetime/

`the uncertain and approximate qualifiers, '?' and '~',  when applied together, are combined into a single qualifier character '%';`

So it seems that you can only have one character to denote the certainty.

**Comment by JeroenDeDauw** [comment]
@mzeinstra so an input such as `1950?~` should be treated as invalid? Or as if it was `1950%` ?

**Comment by mzeinstra** [comment]
Depending how strict we want to be 1950?~ or 1950~? and essentially not valid values, we can automattically transform them to 1950%, but is that expected behaviour by the users of this tool? 

Either it should transform the string or it should throw an error in my opinion. I am leaning a bit to throwing an error as that is more clean than chaning user inputted values.

**Comment by apetushok** [comment]
Thanks @mzeinstra. 
Since there is such support according to the specification https://github.com/ProfessionalWiki/EDTF/pull/100#issuecomment-2073250373
I think it makes sense for us to add it too.
image
cc @JeroenDeDauw

**Comment by mzeinstra** [comment]
> Thanks @mzeinstra. Since there is such support according to the specification [#100 (comment)](https://github.com/ProfessionalWiki/EDTF/pull/100#issuecomment-2073250373) I think it makes sense for us to add it too. image cc @JeroenDeDauw

Not entirely true, there are not examples on that page that combine ? and ~, either ?~ or ~?. So while % means that both the qualifiers ? and ~ apply does not mean that ?~ in a string is valid.

**Comment by apetushok** [comment]
I think you're right. Seems like the best solution here is to throw an error

**Comment by apetushok** [comment]
Fixed 
image

File: src/PackagePrivate/Parser/Parser.php [hunk 1]
```diff
@@ -210,7 +210,7 @@ private function buildQualification(): Qualification {
 	// TODO
 	private static function genQualificationValue( ?string $flag = null ): int {
 		assert( is_string( $flag ) );
-		return (int)self::$map[$flag];
+		return (int)( self::$map[$flag] ?? 0 );
```
[the remark this answers is not in the document]
Comment (apetushok) [reply src/PackagePrivate/Parser/Parser.php:213]: Sure, sorry. I forgot tests.

File: src/PackagePrivate/Parser/Parser.php [hunk 2]
```diff
@@ -209,7 +209,10 @@ private function buildQualification(): Qualification {
 
 	// TODO
 	private static function genQualificationValue( ?string $flag = null ): int {
-		assert( is_string( $flag ) );
+		if ( !is_string( $flag ) || !isset(self::$map[$flag]) ) {
```
Comment (JeroenDeDauw) [inline src/PackagePrivate/Parser/Parser.php:212]: `!array_key_exists( $flag, self::$map )`

File: src/PackagePrivate/Parser/Parser.php [hunk 3]
```diff
@@ -209,7 +209,10 @@ private function buildQualification(): Qualification {
...
+			throw new InvalidArgumentException( 'Qualifier is incorrect' );
```
Comment (JeroenDeDauw) [inline src/PackagePrivate/Parser/Parser.php:213]: `Qualification is invalid`

Comment (JeroenDeDauw) [inline src/PackagePrivate/Parser/Parser.php:213]: `Qualifier is invalid`

File: tests/Unit/PackagePrivate/Parser/ParserTest.php [hunk 4]
```diff
@@ -292,4 +292,27 @@ public function setOpenMiddleNonExtDateValues(): array {
...
+	 * @dataProvider setCombinedUncertainAndApproximateQualifiers
+	 */
+	public function testThrowExceptionWhenUsedCombinedUncertainAndApproximateQualifiers( string $combinedUncertainAndApproximate ): void {
+		$parser = new Parser();
+		$this->expectException( InvalidArgumentException::class );
+		$this->expectExceptionMessage( 'Invalid edtf format ' . $combinedUncertainAndApproximate );
+		$parser->createEdtf( $combinedUncertainAndApproximate );
+	}
+
+	public function setCombinedUncertainAndApproximateQualifiers(): array {
```
Comment (JeroenDeDauw) [inline tests/Unit/PackagePrivate/Parser/ParserTest.php:306]: `provideCombined...`

**Pull request merged**